NIFI-9036 - Add AWS MSK IAM support to Kafka processors

nrnalamalpu opened the pull request:
<!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Enhance Kafka processors for AWS_MSK_IAM SASL mechanism

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

mattyb149 commented:
Is there a Jira case associated with this? If not please create one and use the `NIFI-XXXX` case number in the title of your commit(s), thanks in advance!

nrnalamalpu commented:
Hi @mattyb149, can you please look into this PR?

nrnalamalpu commented:
@pvillard31 or @mattyb149 can anyone take a look at this PR and see if we can merge this to main?

pvillard31 commented:
Just looking at the code, it LGTM. I don't have time to give this a try right now though. Hopefully some time next week if no one else does before.

nrnalamalpu commented:
@pvillard31 or @mattyb149 can we please take a look at this PR and close this out? All checks have been passed for this branch.

nrnalamalpu commented:
@mattyb149 or @pvillard31, Hope you are doing great. Can anyone review and approve this PR ? its been in this state for quite sometime.

nrnalamalpu commented:
@pvillard31, I'm looking into the conflicts. will resolve in a bit and let you guys know. Thanks

pvillard31 commented:
I will, later today

thenatog commented:
@nrnalamalpu Hi, I notice that AwsKmsSensitivePropertyProvider.java is being deleted in this PR - is there any reason this is being removed or is it a mistake? We can't remove this class, and it seems unrelated to the changes you're making. Thanks!

nrnalamalpu commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/KafkaProcessorUtils.java:
```diff
@@ -183,6 +186,32 @@
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .build();
```
Agreed and updated the code accordingly.

nrnalamalpu commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/KafkaProcessorUtils.java:
```diff
@@ -183,6 +186,35 @@
+            .description("specifies the session name that this particular client should use while assuming the IAM role. If the same IAM Role "
+                + "is used in multiple contexts, the session names can be used to differentiate between the different contexts. "
+                + "The awsRoleSessionName is optional")
+            .dependsOn(SASL_MECHANISM, SASL_MECHANISM_AWS_MSK_IAM)
+            .required(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+            .build();
```
I think this property is added as part of new version 1.1.1  of aws-msk-iam-auth. Thank you for looking at the module and verifying it. Added this new property as suggested.

nrnalamalpu commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/KafkaProcessorUtils.java:
```diff
@@ -183,6 +186,35 @@
+    static final PropertyDescriptor AWS_ROLE_ARN = new PropertyDescriptor.Builder()
+            .name("aws.arn")
+            .displayName("AWS Role ARN")
+            .description("The awsRoleArn specifies the ARN for the IAM role the client should use.")
```
Updated

nrnalamalpu commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/KafkaProcessorUtils.java:
```diff
@@ -183,6 +186,35 @@
             .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
             .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
             .build();
+    static final PropertyDescriptor AWS_PROFILE_NAME = new PropertyDescriptor.Builder()
```
@pvillard31 This is a really good observation. But, we don't need to add these properties to each of the classes like ConsumeKafka 2.6, ConsumeKafkaRecord 2.6, PublishKafka 2.6 and PublishKafkaRecord 2.6 if we define them as public static variables. 

**KafkaProcessorUtils.buildCommonKafkaProperties()** gets called from all the above classes which adds all these public static properties to these processors.

Call Stack from Consume/publishKafka classes: 
KafkaProcessorUtils.buildCommonKafkaProperties() --> isStaticStringFieldNamePresent() --> KafkaProcessorUtils.getPublicStaticStringFieldValues()

I tested this within our NIFI cluster and I'll try to attach the screenshot here.

nrnalamalpu commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/KafkaProcessorUtils.java:
```diff
@@ -611,6 +642,29 @@ private static void setScramJaasConfig(final Map<String, Object> mapToPopulate,
+        final String awsRoleArn = context.getProperty(AWS_ROLE_ARN).evaluateAttributeExpressions().getValue();
+        final String awsSessionName = context.getProperty(AWS_SESSION_NAME).evaluateAttributeExpressions().getValue();
+
+        final StringBuilder builder = new StringBuilder("software.amazon.msk.auth.iam.IAMLoginModule required ");
```
We tested the functionality by adding the jar file in the classpath. But, I agree and add the pom dependency.

nrnalamalpu commented:
Hi @pvillard31, 

I could not work on this PR till now due to other priorities. I addressed all the review comments and added the AwsKmsSensitivePropertyProvider.java back which is deleted by mistake.

Please see the screenshots for the testing we did in our AWS environment. 

![image004](https://user-images.githubusercontent.com/88334401/152080858-880c2d36-9c10-442a-9704-f96ff67154e7.jpg)
![image003](https://user-images.githubusercontent.com/88334401/152080793-6af07e0c-b945-4dd2-b8ae-1cc5db798f8e.jpg)

nrnalamalpu commented:
@pvillard31, Could you please look at the changes and see if we can get this merged to main branch? Thanks

nrnalamalpu commented:
Can anyone please look into this PR and see if we can close this out?

thanks

nandorsoma reviewed (commented):
Hey @nrnalamalpu!
I've reviewed this PR. It wasn't easy because the git history is a bit messy, which makes it hard to apply to the current main. What I see is that the new PropertyDescriptors are not registered in the processors. Without that, this feature won't work. Did you leave it out from one of the commits by mistake?
I see that this PR has been open for a long time. Do you have the intention to wrap it up?

nrnalamalpu commented:
Hi @nandorsoma, Glad you reviewed the pull request at least. I was waiting for at least one of the committers to review and approve the PR. I would like to work on this and wrap up. Please let me know what should I do to wrap this up.

nandorsoma commented:
Hey @nrnalamalpu!
As I mentioned in the previous comment, the PropertyDescriptors are not registered in any of the processors so this feature won't work currently. The first step would be to fix that. After that:
- Rebase the branch on top of the current main.
- Since the git history is a bit messy, it would be hard to merge it. Please squash all of your commits and force-push one single commit, which can be our new baseline. If it's easier for you, you can open a new PR and do a cross-reference (commenting the link of the PR is enough) on both PR.
- Test it with MSK, and only update this PR when it works as expected.

nandorsoma commented:
@nrnalamalpu any update on this?

nandorsoma commented:
Hey @exceptionfactory! Thank you for the heads up! I'm interested in opening a new pr, but I'd like to wait for @nrnalamalpu's answer.

nrnalamalpu commented:
Hi @nandorsoma and @exceptionfactory, we can have a new PR. Unfortunately, I tied up with other Priority items and could not spend time on this.

nrnalamalpu commented:
@nandorsoma, not sure which file you are referring for the property descriptors. The PropertyDescriptors are added in KafkaProcessorUtils.java file which will be rolled up to consumeKafka/publishKafka processors.

nandorsoma commented:
@nrnalamalpu, how do you mean they will be rolled up? For example, `AWS_PROFILE_NAME` is referenced only in one place in the whole codebase when you get the value of the property in `setIamJaasConfig()`. If you check other PropertyDescriptors, such as `PASSWORD`, you can see that it is added to the descriptor list of all processors that use it.
